src/main/java/com/google/api/codegen/gapic/GapicProviderFactory.java

```diff
@@ -0,0 +1,202 @@
+/* Copyright 2016 Google Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.api.codegen.gapic;
+
+import com.google.api.codegen.ApiConfig;
+import com.google.api.codegen.InterfaceListView;
+import com.google.api.codegen.InterfaceView;
+import com.google.api.codegen.ProtoFileView;
+import com.google.api.codegen.clientconfig.ClientConfigGapicContext;
+import com.google.api.codegen.clientconfig.ClientConfigSnippetSetRunner;
+import com.google.api.codegen.csharp.CSharpGapicContext;
+import com.google.api.codegen.csharp.CSharpSnippetSetRunner;
+import com.google.api.codegen.go.GoGapicContext;
+import com.google.api.codegen.go.GoSnippetSetRunner;
+import com.google.api.codegen.java.JavaGapicContext;
+import com.google.api.codegen.java.JavaIterableSnippetSetRunner;
+import com.google.api.codegen.java.JavaSnippetSetRunner;
+import com.google.api.codegen.nodejs.NodeJSGapicContext;
+import com.google.api.codegen.nodejs.NodeJSSnippetSetRunner;
+import com.google.api.codegen.php.PhpGapicContext;
+import com.google.api.codegen.php.PhpSnippetSetRunner;
+import com.google.api.codegen.py.PythonGapicContext;
+import com.google.api.codegen.py.PythonInterfaceInitializer;
+import com.google.api.codegen.py.PythonProtoFileInitializer;
+import com.google.api.codegen.py.PythonSnippetSetRunner;
+import com.google.api.codegen.ruby.RubyGapicContext;
+import com.google.api.codegen.ruby.RubySnippetSetRunner;
+import com.google.api.tools.framework.model.Interface;
+import com.google.api.tools.framework.model.Model;
+import com.google.api.tools.framework.model.ProtoFile;
+
+import java.util.Arrays;
+import java.util.List;
+
+import org.apache.commons.lang3.NotImplementedException;
+
+/**
+ * GapicProviderFactory creates GapicProvider instances based on an id.
+ */
+public class GapicProviderFactory {
+
+  /**
+   * Create the GapicProviders based on the given id
+   */
+  public static List<GapicProvider<? extends Object>> defaultCreate(Model model,
+      ApiConfig apiConfig, String id) {
+
+    // Please keep the following IDs in alphabetical order
+
+    if (id.equals("client_config")) {
```

Comment:
Naked strings scattered throughout the code are almost never a good thing (https://en.wikipedia.org/wiki/Magic_number_(programming)#Unnamed_numerical_constants).

Turning these into constants buys us an easy-to-read table of contents for expected values. It also encourages others to use the same constants, rather than hand-writing the same or similar strings (especially any other code that might ever call into this specific method, such as tests).

Having the compiler catch such errors is even better than failing at start time.